File: src/HomeWindow.cpp
#include "HomeWindow.h"

#include <QButtonGroup>
#include <QDir>
#include <QHBoxLayout>
#include <QLabel>
#include <QLineEdit>
#include <QPushButton>
#include <QStandardPaths>
#include <QVBoxLayout>
#include <QWidget>

HomeWindow::HomeWindow(QWidget *parent)
    : QMainWindow(parent) {
  setupUi();
  applyStyles();
}

void HomeWindow::setupUi() {
  auto *central = new QWidget(this);
  setCentralWidget(central);
  resize(980, 680);
  setMinimumSize(840, 560);
  setWindowTitle("SlideApp");

  auto *root = new QVBoxLayout(central);
  root->setContentsMargins(70, 30, 70, 40);
  root->setSpacing(0);

  auto *topRow = new QHBoxLayout();
  topRow->addStretch();

  researcherButton_ = new QPushButton("Researcher Mode", this);
  researcherButton_->setCheckable(true);

  adminButton_ = new QPushButton("Admin Mode", this);
  adminButton_->setCheckable(true);

  auto *modeGroup = new QButtonGroup(this);
  modeGroup->setExclusive(true);
  modeGroup->addButton(researcherButton_);
  modeGroup->addButton(adminButton_);

  topRow->addWidget(researcherButton_);
  topRow->addSpacing(10);
  topRow->addWidget(adminButton_);
  root->addLayout(topRow);

  root->addSpacing(86);

  auto *title = new QLabel("Emerson Lab", this);
  title->setAlignment(Qt::AlignHCenter);
  title->setObjectName("titleLabel");
  root->addWidget(title);

  root->addSpacing(44);

  auto *formWrap = new QWidget(this);
  formWrap->setMaximumWidth(760);
  auto *form = new QVBoxLayout(formWrap);
  form->setContentsMargins(0, 0, 0, 0);
  form->setSpacing(18);

  auto *userRow = new QHBoxLayout();
  userRow->setSpacing(14);
  auto *userLabel = new QLabel("User:", this);
  userLabel->setObjectName("fieldLabel");
  userEdit_ = new QLineEdit(this);
  userEdit_->setObjectName("fieldInput");
  userEdit_->setMinimumWidth(420);
  userEdit_->setPlaceholderText("");
  userRow->addWidget(userLabel);
  userRow->addWidget(userEdit_, 1);
  form->addLayout(userRow);

  auto *outputRow = new QHBoxLayout();
  outputRow->setSpacing(14);
  auto *outputLabel = new QLabel("Output Directory (Optional):", this);
  outputLabel->setObjectName("fieldLabel");
  outputDirEdit_ = new QLineEdit(this);
  outputDirEdit_->setObjectName("fieldInput");
  outputDirEdit_->setMinimumWidth(420);
  outputDirEdit_->setPlaceholderText("");

  QString downloads = QStandardPaths::writableLocation(QStandardPaths::DownloadLocation);
  if (downloads.isEmpty()) {
    downloads = QDir::homePath() + "/Downloads";
  }
  outputDirEdit_->setText(downloads);

  outputRow->addWidget(outputLabel);
  outputRow->addWidget(outputDirEdit_, 1);
  form->addLayout(outputRow);

  auto *formCenterRow = new QHBoxLayout();
  formCenterRow->addStretch();
  formCenterRow->addWidget(formWrap);
  formCenterRow->addStretch();
  root->addLayout(formCenterRow);

  root->addSpacing(64);

  beginButton_ = new QPushButton("Begin", this);
  beginButton_->setObjectName("beginButton");
  beginButton_->setFixedSize(210, 74);

  auto *beginRow = new QHBoxLayout();
  beginRow->addStretch();
  beginRow->addWidget(beginButton_);
  beginRow->addStretch();
  root->addLayout(beginRow);

  root->addStretch();

  researcherButton_->setChecked(true);
  setModeResearcher();

  connect(researcherButton_, &QPushButton::clicked, this, &HomeWindow::setModeResearcher);
  connect(adminButton_, &QPushButton::clicked, this, &HomeWindow::setModeAdmin);
}

void HomeWindow::applyStyles() {
  setStyleSheet(
      "QMainWindow { background: #000000; }"
      "QWidget { color: #f2f2f2; font-family: 'Helvetica Neue'; }"
      "QLabel#titleLabel { font-size: 78px; font-weight: 300; letter-spacing: 1px; }"
      "QLabel#fieldLabel { font-size: 42px; font-weight: 400; }"
      "QLineEdit#fieldInput {"
      "  background: transparent;"
      "  border: none;"
      "  color: #ffffff;"
      "  font-size: 42px;"
      "  font-weight: 400;"
      "  padding: 0;"
      "}"
      "QPushButton {"
      "  background: #d9d9d9;"
      "  color: #111111;"
      "  border: none;"
      "  border-radius: 11px;"
      "  padding: 10px 18px;"
      "  font-size: 25px;"
      "}"
      "QPushButton:hover { background: #ececec; }"
      "QPushButton:pressed { background: #bbbbbb; }"
      "QPushButton#beginButton {"
      "  background: #84F28A;"
      "  color: #000000;"
      "  border-radius: 36px;"
      "  font-size: 58px;"
      "  font-weight: 500;"
      "  padding-bottom: 8px;"
      "}"
      "QPushButton#beginButton:hover { background: #95f69a; }"
      "QPushButton#beginButton:pressed { background: #76e87d; }");
}

void HomeWindow::setModeResearcher() {
  researcherButton_->setChecked(true);

  researcherButton_->setStyleSheet(
      "QPushButton {"
      "  background: #84F28A;"
      "  color: #000000;"
      "  border: none;"
      "  border-radius: 11px;"
      "  padding: 10px 18px;"
      "  font-size: 25px;"
      "}");

  adminButton_->setStyleSheet(
      "QPushButton {"
      "  background: #d9d9d9;"
      "  color: #111111;"
      "  border: none;"
      "  border-radius: 11px;"
      "  padding: 10px 18px;"
      "  font-size: 25px;"
      "}");
}

void HomeWindow::setModeAdmin() {
  adminButton_->setChecked(true);

  adminButton_->setStyleSheet(
      "QPushButton {"
      "  background: #F26565;"
      "  color: #000000;"
      "  border: none;"
      "  border-radius: 11px;"
      "  padding: 10px 18px;"
      "  font-size: 25px;"
      "}");

  researcherButton_->setStyleSheet(
      "QPushButton {"
      "  background: #d9d9d9;"
      "  color: #111111;"
      "  border: none;"
      "  border-radius: 11px;"
      "  padding: 10px 18px;"
      "  font-size: 25px;"
      "}");
}

File: pyapp/home_window.py
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import (
    QCheckBox,
    QButtonGroup,
    QHBoxLayout,
    QLabel,
    QLineEdit,
    QMainWindow,
    QPushButton,
    QVBoxLayout,
    QWidget,
)

from pyapp.database import verify_credentials


class HomeWindow(QMainWindow):
    login_success = Signal(str, str)

    def __init__(self) -> None:
        super().__init__()
        self.researcher_button: QPushButton | None = None
        self.admin_button: QPushButton | None = None
        self.user_edit: QLineEdit | None = None
        self.password_edit: QLineEdit | None = None
        self.begin_button: QPushButton | None = None
        self.status_label: QLabel | None = None

        self._build_ui()
        self._apply_styles()
        self._set_researcher_mode()

    def _build_ui(self) -> None:
        self.setWindowTitle("SlideApp")
        self.resize(980, 680)
        self.setMinimumSize(840, 560)

        central = QWidget(self)
        self.setCentralWidget(central)

        root = QVBoxLayout(central)
        root.setContentsMargins(70, 30, 70, 40)
        root.setSpacing(0)

        top_row = QHBoxLayout()
        top_row.addStretch()

        self.researcher_button = QPushButton("Researcher Mode", self)
        self.researcher_button.setCheckable(True)

        self.admin_button = QPushButton("Admin Mode", self)
        self.admin_button.setCheckable(True)

        mode_group = QButtonGroup(self)
        mode_group.setExclusive(True)
        mode_group.addButton(self.researcher_button)
        mode_group.addButton(self.admin_button)

        top_row.addWidget(self.researcher_button)
        top_row.addSpacing(10)
        top_row.addWidget(self.admin_button)
        root.addLayout(top_row)

        root.addSpacing(86)

        title = QLabel("Emerson Lab", self)
        title.setAlignment(Qt.AlignmentFlag.AlignHCenter)
        title.setObjectName("titleLabel")
        root.addWidget(title)

        root.addSpacing(44)

        form_wrap = QWidget(self)
        form_wrap.setMaximumWidth(760)
        form = QVBoxLayout(form_wrap)
        form.setContentsMargins(0, 0, 0, 0)
        form.setSpacing(18)

        user_row = QHBoxLayout()
        user_row.setSpacing(14)
        user_label = QLabel("User:", self)
        user_label.setObjectName("fieldLabel")

        self.user_edit = QLineEdit(self)
        self.user_edit.setObjectName("fieldInput")
        self.user_edit.setMinimumWidth(420)
        self.user_edit.setPlaceholderText("")

        user_row.addWidget(user_label)
        user_row.addWidget(self.user_edit, 1)
        form.addLayout(user_row)

        password_row = QHBoxLayout()
        password_row.setSpacing(14)
        password_label = QLabel("Password:", self)
        password_label.setObjectName("fieldLabel")

        self.password_edit = QLineEdit(self)
        self.password_edit.setObjectName("fieldInput")
        self.password_edit.setMinimumWidth(420)
        self.password_edit.setPlaceholderText("")
        self.password_edit.setEchoMode(QLineEdit.EchoMode.Password)

        show_password_checkbox = QCheckBox("Show", self)
        show_password_checkbox.setObjectName("showPasswordCheck")
        show_password_checkbox.toggled.connect(
            lambda checked: self.password_edit.setEchoMode(
                QLineEdit.EchoMode.Normal if checked else QLineEdit.EchoMode.Password
            )
        )

        password_row.addWidget(password_label)
        password_row.addWidget(self.password_edit, 1)
        password_row.addWidget(show_password_checkbox)
        form.addLayout(password_row)

        self.status_label = QLabel("", self)
        self.status_label.setObjectName("statusLabel")
        form.addWidget(self.status_label)

        form_center_row = QHBoxLayout()
        form_center_row.addStretch()
        form_center_row.addWidget(form_wrap)
        form_center_row.addStretch()
        root.addLayout(form_center_row)

        root.addSpacing(64)

        self.begin_button = QPushButton("Begin", self)
        self.begin_button.setObjectName("beginButton")
        self.begin_button.setFixedSize(280, 110)

        begin_row = QHBoxLayout()
        begin_row.addStretch()
        begin_row.addWidget(self.begin_button)
        begin_row.addStretch()
        root.addLayout(begin_row)

        root.addStretch()

        self.researcher_button.clicked.connect(self._set_researcher_mode)
        self.admin_button.clicked.connect(self._set_admin_mode)
        self.begin_button.clicked.connect(self._handle_begin_clicked)

    def _apply_styles(self) -> None:
        self.setStyleSheet(
            "QMainWindow { background: #000000; }"
            "QWidget { color: #f2f2f2; font-family: 'Helvetica Neue'; }"
            "QLabel#titleLabel { font-size: 78px; font-weight: 300; letter-spacing: 1px; }"
            "QLabel#fieldLabel { font-size: 42px; font-weight: 400; }"
            "QLineEdit#fieldInput {"
            "  background: transparent;"
            "  border: none;"
            "  color: #ffffff;"
            "  font-size: 42px;"
            "  font-weight: 400;"
            "  padding: 0;"
            "}"
            "QPushButton {"
            "  background: #d9d9d9;"
            "  color: #111111;"
            "  border: none;"
            "  border-radius: 11px;"
            "  padding: 10px 18px;"
            "  font-size: 25px;"
            "}"
            "QPushButton:hover { background: #ececec; }"
            "QPushButton:pressed { background: #bbbbbb; }"
            "QPushButton#beginButton {"
            "  background: #84F28A;"
            "  color: #000000;"
            "  border-radius: 48px;"
            "  font-size: 64px;"
            "  font-weight: 500;"
            "  padding: 0;"
            "}"
            "QPushButton#beginButton:hover { background: #95f69a; }"
            "QPushButton#beginButton:pressed { background: #76e87d; }"
            "QLabel#statusLabel { color: #ff9a9a; font-size: 20px; min-height: 24px; }"
            "QCheckBox#showPasswordCheck { color: #d4d4d4; font-size: 22px; }"
        )

    def _set_researcher_mode(self) -> None:
        if self.researcher_button is None or self.admin_button is None:
            return

        self.researcher_button.setChecked(True)
        self.researcher_button.setStyleSheet(
            "QPushButton {"
            "  background: #84F28A;"
            "  color: #000000;"
            "  border: none;"
            "  border-radius: 11px;"
            "  padding: 10px 18px;"
            "  font-size: 25px;"
            "}"
        )

        self.admin_button.setStyleSheet(
            "QPushButton {"
            "  background: #d9d9d9;"
            "  color: #111111;"
            "  border: none;"
            "  border-radius: 11px;"
            "  padding: 10px 18px;"
            "  font-size: 25px;"
            "}"
        )

    def _set_admin_mode(self) -> None:
        if self.researcher_button is None or self.admin_button is None:
            return

        self.admin_button.setChecked(True)
        self.admin_button.setStyleSheet(
            "QPushButton {"
            "  background: #F26565;"
            "  color: #000000;"
            "  border: none;"
            "  border-radius: 11px;"
            "  padding: 10px 18px;"
            "  font-size: 25px;"
            "}"
        )

        self.researcher_button.setStyleSheet(
            "QPushButton {"
            "  background: #d9d9d9;"
            "  color: #111111;"
            "  border: none;"
            "  border-radius: 11px;"
            "  padding: 10px 18px;"
            "  font-size: 25px;"
            "}"
        )

    def _handle_begin_clicked(self) -> None:
        if (
            self.user_edit is None
            or self.password_edit is None
            or self.status_label is None
        ):
            return

        username = self.user_edit.text().strip()
        password = self.password_edit.text()
        if not username or not password:
            self.status_label.setText("Enter username and password.")
            return

        role = verify_credentials(username, password)
        if role is None:
            self.status_label.setText("Invalid credentials.")
            return

        if self.admin_button is not None and self.admin_button.isChecked() and role != "admin":
            self.status_label.setText("Admin mode requires an admin account.")
            return

        mode = "admin" if self.admin_button is not None and self.admin_button.isChecked() else "researcher"
        self.status_label.setText("")
        self.login_success.emit(username, mode)
